docs/userguide/design.md

```diff
@@ -206,35 +206,84 @@ service objects, organized according to locale. Then, if a particular locale's
 formatter is in high demand, that formatter can be used, and then returned to
 the pool.
 
-#### ICU Memory Usage
+### ICU Memory Usage
 
 ICU4C APIs are designed to allow separate heaps for its libraries vs. the
 application. This is achieved by providing functions to allocate and release
 objects owned by ICU4C using only ICU4C library functions. For more details see
-the Memory Usage section in the [Coding Guidelines](dev/codingguidelines.md).
+the Memory Usage section in the [Coding Guidelines](dev/codingguidelines.md#memory-usage).
 
-#### ICU Initialization and Termination
+### ICU4C Initialization and Termination
 
 The ICU library does not normally require any explicit initialization prior to
 use. An application begins use simply by calling any ICU API in the usual way.
-(There is one exception to this, described below.)
+There are, however, a few functions affecting ICU's configuration, that, if used,
+must be called first, before other use of ICU in a process. These are outlined below.
 
-In C++ programs, ICU objects and APIs may safely be used during static
-initialization of other application-defined classes or objects. There are no
-order-of-initialization problems between ICU and static objects from other
-libraries because ICU does not rely on C++ static object initialization for its
-normal operation.
+1. `u_setMemoryFunctions()`. This function replaces the standard library heap
+allocation functions used by ICU with alternate versions, provided by the
+application. If it is needed, `u_setMemoryFunctions()` must be called first, before
+any other use of ICU. This functionality is not commonly used.
+
+2. ICU Data Locating Functions, `u_setCommonData()`, `u_setDataDirectory()`, and
+`u_setAppData()`. One or more of these functions will be required when ICU is
+configured to load its data directly from files rather than taking it from the
+default data dll, and the files are not in the default location. Again, this is
+not common. See [ICU Data](icudata#icu-data-directory).
+
+3. `u_init()`. This function performs a sanity check on ICU's ability to access
```

Comment:
This is all true, but isn't u_init() down to loading a single data file (convrtrs.txt or something)?
As a "sanity check", should we suggest trying to call something that the program will call later anyway, and that will require data, such as formatting a number or a date?

Reply:
wfm

Reply:
> The existing u_init() does a ucnv_io_countKnownConverters(). So what happens if ICU is configured with no converters?

FWIW, we actually ran into this issue with ICU when it was built using a data file instead of a data DLL. :)

IIRC, when there were no converters to load, the `u_init()` call effectively became a no-op, and so it never actually caused the ICU data file to be memory mapped and loaded at all, and would just return success. This mean that later on other APIs would fail with errors, as they would internally cause the data file to be loaded, and possibly hit out-of-memory, or otherwise fail to load the data file. Or worse, since not all ICU APIs have `UErrorCode` parameters it meant that sometimes APIs would randomly return garbage results... This was all despite initially calling `u_init()` and having it return *success*.

What we ended up doing instead was to call `ulocdata_getCLDRVersion()` first before calling other ICU APIs.

I chose this API instead for the following reasons:
- While some people may choose to omit the converters (due to size, or perhaps not needing them) when building a custom ICU data file, the CLDR version number takes up very little space in the data, and so is generally never removed or filtered out of the data.
- It's exported from the icui18n library, so being able to successfully call it means that both icui18n and icuuc have been loaded. Whereas `u_init` is exported from the icuuc library -- meaning that icui18n might not be loaded yet, and so you can possibly run into OOM issues later on when icui18n is loaded.
- The amount of work `ulocdata_getCLDRVersion` does is fairly trivial compared to other icui18n APIs.

FWIW, you can see an example of this in ChakraCore here:
https://github.com/microsoft/ChakraCore/blob/861a276c346746f3b87347bb094988563d7ee0a7/lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp#L610
.NET Core also does the same here:
https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L369

On one hand, I don't know how common it is to use a data file + omit converters. Though on the other hand, it might be nice to add a small note maybe, as this has bitten at least a few.